### ISSUE TITLE
Extension API: Ability to cancel/stop redundant threads for the event listeners

### ISSUE THREAD

**Issue by sunn4room** [Ulauncher/Ulauncher#1112]
### Description

When I type something to get result items, the query operation may be a time-consuming operation. There is a simple extension:
```python
class DemoExtension(Extension):
    def on_input(self, input_text, trigger_id):
        self.logger.debug(input_text)
        time.sleep(2) # Simulate time-consuming operation
        self.logger.debug(input_text)
        for i in range(5):
            yield ExtensionResult(
                name='Item %s' % i,
                description='Item description %s' % i,
                on_enter=HideWindowAction()
            )
```
Type 1234 continuously(regardless of input debounce), and the debug message is:
```
1
(2 seconds)
12
(2 seconds)
123
(2 seconds)
1234
```
So I need a new feature about non-blocking API. These is my proposal:
```python
def on_input(self, input_text, trigger_id, set_result_items):
    # set_result_items is a function to set ResultItem List
    ... # terminate the last query process
    ... # create a new process, and set result items with that function
    # this function can return immediately
```
I welcome any feedback or suggestions now.

### Communication guidelines

- [X] I have read and agree with Ulauncher's communication guidelines and checked the [existing feature requests](https://github.com/Ulauncher/Ulauncher/labels/feature) before suggesting this.

### Feature isn't possible to support as an extension.

- [X] This feature request cannot satisfactory be supported as a third party extension.

**Comment by friday** [comment]
I really appreciate your feedbacks and suggestions. I mostly avoided slow extensions myself, so I didn't see the extent of the issue. However, I think the solution isn't needed. What do you think of my counter-suggestion? ^

**Comment by sunn4room** [comment]
I agree your counter-suggestion because this is a project written in python. Compared to javascript, python is not good at asynchronous programming and GUI programming although python has asyncio, threading and so on. Blocking problems are inevitable.

But I don't think that you can avoided slow extensions everytime. `https://github.com/brpaz/ulauncher-file-search` uses `fd` to search file. `fd` aften take a few secounds although it is fast enough.  Fortunately, a remedy here is to increase the input debounce.

This is a whimsical issue. Extensibility and non-blocking are contradictory. So I close this issue.

**Comment by friday** [comment]
> Compared to javascript, python is not good at asynchronous programming and GUI programming.

I agree. I think JavaScript is actually way better for what we do (especially Typescript) because GTK has first class JS support, it's overall faster, more functional and doesn't require any specific runtime. You can build all your third party dependencies to one file, rather than all your users having to install them too. Although if extensions were integrated with deno they wouldn't have to.

> Although python has asyncio, threading and so on. Blocking problems are inevitable.

I don't know if I necessary agree with this. You can work around it like I did and threading actually utilizes multiple CPU cores better. In JS it's single threaded only, but it's easier to write non-blocking code and that mostly solves the real performance issues.

> But I don't think that you can avoided slow extensions everytime. `https://github.com/brpaz/ulauncher-file-search` uses `fd` to search file. `fd` aften take a few secounds although it is fast enough. Fortunately, a remedy here is to increase the input debounce.

Yeah. You probably don't want to use an extension like this that actually searches the system, rather than caching the system so it can search it later. locate and gnome tracker is much faster and easier on your system resources.

> This is a whimsical issue. Extensibility and non-blocking are contradictory. So I close this issue.

No. Your issues are all very well founded, and helped make Ulauncher better 👍 But I guess we can consider this solved.

**Comment by sunn4room** [comment]
Maybe it's a good idea to make a new thread.
```python
# Client.py: 56
-            self.extension.trigger_event(event)
+            def extension_trigger_event():
+                self.extension.trigger_event(event)
+            threading.Thread(target=extension_trigger_event).start()
```
it can avoid blocking the main thread. And it works fine with my time-consuming extension. But I don't know what the side effects will be, because I haven't understand the total source code about event-listener invocation.

I think the reason for the time-consuming extension is mainly the call of external shell command, so I have a new proposal:
```python
# extension.py
...
def on_input(self, input_text, id):
    return subprocess.Popen([...], stdout..., stderr...)

def on_popen_output(self, output):
    # normal return
```
Next, I will try to complete this proposal and make a push.

I will reopen this issue, because I think this issue deserves further discussion

**Comment by friday** [comment]
> it can avoid blocking the main thread. And it works fine with my time-consuming extension. But I don't know what the side effects will be, because I haven't understand the total source code about event-listener invocation.

It should be fine. The only problem I see is if you want it to be blocking. If you for example need to write to make transactions in the right order to a database. Or if you need to do something slow and blocking after the user changes the settings. But it's better to default to this, and people can handle the niche cases.

> # extension.py
> ...
> def on_input(self, input_text, id):
>     return subprocess.Popen([...], stdout..., stderr...)
> 
> def on_popen_output(self, output):
>     # normal return

Why? This doesn't solve anything, and it makes it more complex. subprocess is only one type of blocking operation. There anre many others.

**Comment by friday** [comment]
I'm going to merge the PR. The way I see it it fully solves the problem and doesn't add any cost. It could potentially be improved on, like adding the ability to cancel/invalidate threads when you type the next letter, but I don't think you can do this with threading and if you go for something other than threading you use a single thread, and don't get the benefit of utilizing multiple cores.

If you have a counter-suggestion you're welcome to make a PR, but I think you're over-thinking and should try to understand the current solution and look at the python concurrency models (and remember it has to work for python 3.6) before suggestion semantics from other languages.

**Comment by sunn4room** [comment]
> Why? This doesn't solve anything, and it makes it more complex. subprocess is only one type of blocking operation. There are many others. And there are different methods of subprocess. And there are third party python libraries that wrap commands, that doesn't let you access the subprocess directly.

I just want the extension to have the ability to handle external command output, including the termination of command when you type the next letter.  `subprocess` isn't matter, and the output of command is matter. Too many extensions in the community use the output of external command to render result items. That's all.

The new PR is great enough for me. And it's better to add `threading.Event` for graceful exit. Good job.

**Comment by friday** [comment]
Ok :+1: I renamed the title to clarify/limit the scope a bit.

**Comment by friday** [comment]
I think there are no more actions to be taken, so closing this (4 years after).
